chapter6_marl/marl_network.py

```diff
@@ -0,0 +1,244 @@
+
+
+class ActorCriticNetwork(nn.Module):
+
+    def __init__(self, obs_shape: int, action_shape: int) -> None:
+        """
+        **Overview**:
+            The definition of basic actor-critic network in policy gradient algorithms (e.g. PG/A2C/PPO), which is mainly composed of
+            three parts: encoder, policy head and value head.
+        """
+        # PyTorch necessary requirements for extending ``nn.Module`` . Our network should also subclass this class.
+        super(ActorCriticNetwork, self).__init__()
+        # Define encoder module, which maps raw local state of each agent into embedding vector.
+        # It could be different for various state, such as Convolution Neural Network (CNN) for image state and Multilayer perceptron (MLP) for vector state, respectively.
+        # Here we use two-layer MLP for vector state, i.e.
+        # $$y = max(W_2 max(W_1 x+b_1, 0) + b_2, 0)$$
+        self.encoder = nn.Sequential(
+            nn.Linear(obs_shape, 32),
+            nn.ReLU(),
+            nn.Linear(32, 64),
+            nn.ReLU(),
+        )
+        # Define discrete action logit output network, just one-layer FC.
+        self.policy_head = nn.Linear(64, action_shape)
+        # Define scalar value output network.
+        self.value_head = nn.Linear(64, 1)
+
+    # delimiter
+    def forward(self, local_obs: torch.Tensor) -> torch.Tensor:
+        """
+        **Overview**:
+            The computation graph of actor-critic network in discrete action space.
+        """
+        # Transform original local obs into embedding vector, i.e. $$(B, A, *) -> (B, A, N)$$
+        # Some network layers in PyTorch like ``nn.Linear`` can deal with any number of prefix dimensions, so we can just use it to process the whole multi-agent batch.
+        x = self.encoder(local_obs)
+        # Calculate logit for each possible discrete action choices, i.e. $$(B, A, N) -> (B, A, M)$$
+        logit = self.policy_head(x)
+        # Calculate value for each sample and agent, i.e. $$(B, A, N) -> (B, A, 1)$$
+        value = self.value_head(x)
+        # Return the final result by treetensor format.
+        return ttorch.as_tensor({
+            'logit': logit,
+            'value': value,
+        })
+
+
+# delimiter
+class SharedActorCriticNetwork(nn.Module):
+
+    def __init__(self, agent_num: int, obs_shape: int, action_shape: int) -> None:
+        """
+        **Overview**:
+            The definition of shared parameters actor-critic network in policy gradient algorithms for multi-agent scenarios.
+            Each agent shares the same parameters in the network so that they can be processed as a batch in parallel.
+        """
+        # PyTorch necessary requirements for extending ``nn.Module`` . Our network should also subclass this class.
+        super(SharedActorCriticNetwork, self).__init__()
+        # Define a shared actor-critic network used for all the agents.
+        self.agent_num = agent_num
```

Comment:
The agent_num defined here seems to have no effect. Perhaps it can be used to illustrate the above-mentioned "Each agent shares the same parameters in the network so that they can be processed as a batch in parallel.". For example, the shape of input is (batch_size, agent_num, obs_shape)